### PR TITLE
[harfbuzz] Remove PARENT_SCOPE in harfbuzzConfig.cmake.in

### LINKED ISSUE #23694: [harfbuzz]  Regression: Cannot set "HARFBUZZ_LIBRARY": current scope has no parent
**Describe the bug**
With the latest version of vcpkg, when I install harfbuzz and use `find_package(harfbuzz)`  (or  `find_package(harfbuzz CONFIG)`) in a CMakeLists.txt, I get the following warnings (full warning messages below):

```
Cannot set "HARFBUZZ_LIBRARY": current scope has no parent.
Cannot set "HARFBUZZ_LIBRARIES": current scope has no parent.
Cannot set "HARFBUZZ_INCLUDE_DIR": current scope has no parent.
Cannot set "HARFBUZZ_INCLUDE_DIRS": current scope has no parent.
Cannot set "HARFBUZZ_FOUND": current scope has no parent.
```

and indeed, if we print those variable to check, we can see that the variables are not set.

**Environment**
- OS: Windows 11
- Compiler: MSVC 2019
- CMake 3.22.2 or 3.23.0

**To Reproduce**
1. ./vcpkg install harfbuzz
2. Create the following CMakeLists.txt:

```
cmake_minimum_required(VERSION 3.23)
project(harfbuzz-warning)
find_package(harfbuzz)
message(STATUS "HARFBUZZ_LIBRARY      = ${HARFBUZZ_LIBRARY}")
message(STATUS "HARFBUZZ_LIBRARIES    = ${HARFBUZZ_LIBRARIES}")
message(STATUS "HARFBUZZ_INCLUDE_DIR  = ${HARFBUZZ_INCLUDE_DIR}")
message(STATUS "HARFBUZZ_INCLUDE_DIRS = ${HARFBUZZ_INCLUDE_DIRS}")
message(STATUS "HARFBUZZ_FOUND        = ${HARFBUZZ_FOUND}")
```

3. cmake .. -DCMAKE_TOOLCHAIN_FILE="$HOME\vcpkg\scripts\buildsystems\vcpkg.cmake":

```
PS C:\Users\dalbo\test\harfbuzz-warning\build> cmake .. -DCMAKE_TOOLCHAIN_FILE="$HOME\vcpkg\scripts\buildsystems\vcpkg.cmake"
-- Building for: Visual Studio 16 2019
-- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.22000.
-- The C compiler identification is MSVC 19.29.30140.0
-- The CXX compiler identification is MSVC 19.29.30140.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found ZLIB: optimized;C:/Users/dalbo/vcpkg/installed/x64-windows/lib/zlib.lib;debug;C:/Users/dalbo/vcpkg/installed/x64-windows/debug/lib/zlibd.lib (found version "1.2.11")
CMake Warning (dev) at C:/Users/dalbo/vcpkg/installed/x64-windows/share/harfbuzz/harfbuzzConfig.cmake:62 (set):
  Cannot set "HARFBUZZ_LIBRARY": current scope has no parent.
Call Stack (most recent call first):
  C:/Users/dalbo/vcpkg/scripts/buildsystems/vcpkg.cmake:782 (_find_package)
  CMakeLists.txt:3 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at C:/Users/dalbo/vcpkg/installed/x64-windows/share/harfbuzz/harfbuzzConfig.cmake:63 (set):
  Cannot set "HARFBUZZ_LIBRARIES": current scope has no parent.
Call Stack (most recent call first):
  C:/Users/dalbo/vcpkg/scripts/buildsystems/vcpkg.cmake:782 (_find_package)
  CMakeLists.txt:3 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at C:/Users/dalbo/vcpkg/installed/x64-windows/share/harfbuzz/harfbuzzConfig.cmake:64 (set):
  Cannot set "HARFBUZZ_INCLUDE_DIR": current scope has no parent.
Call Stack (most recent call first):
  C:/Users/dalbo/vcpkg/scripts/buildsystems/vcpkg.cmake:782 (_find_package)
  CMakeLists.txt:3 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at C:/Users/dalbo/vcpkg/installed/x64-windows/share/harfbuzz/harfbuzzConfig.cmake:65 (set):
  Cannot set "HARFBUZZ_INCLUDE_DIRS": current scope has no parent.
Call Stack (most recent call first):
  C:/Users/dalbo/vcpkg/scripts/buildsystems/vcpkg.cmake:782 (_find_package)
  CMakeLists.txt:3 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at C:/Users/dalbo/vcpkg/installed/x64-windows/share/harfbuzz/harfbuzzConfig.cmake:68 (set):
  Cannot set "HARFBUZZ_FOUND": current scope has no parent.
Call Stack (most recent call first):
  C:/Users/dalbo/vcpkg/scripts/buildsystems/vcpkg.cmake:782 (_find_package)
  CMakeLists.txt:3 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- HARFBUZZ_LIBRARY      =
-- HARFBUZZ_LIBRARIES    =
-- HARFBUZZ_INCLUDE_DIR  =
-- HARFBUZZ_INCLUDE_DIRS =
-- HARFBUZZ_FOUND        =
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/dalbo/test/harfbuzz-warning/build
```

**Expected behavior**
There should be no warning and the variables should be set.

**How to fix?**

The offending lines of code are [L62-71 of vcpkg/ports/harfbuzz/harfbuzzConfig.cmake.in ](https://github.com/microsoft/vcpkg/blob/a3df696929b28195b0d7a477ba6cd16dba48fa52/ports/harfbuzz/harfbuzzConfig.cmake.in#L62):

```
set(HARFBUZZ_LIBRARY harfbuzz::harfbuzz PARENT_SCOPE)
set(HARFBUZZ_LIBRARIES harfbuzz::harfbuzz PARENT_SCOPE)
set(HARFBUZZ_INCLUDE_DIR "${_INSTALL_DIR}/include/harfbuzz" PARENT_SCOPE)
set(HARFBUZZ_INCLUDE_DIRS "${_INSTALL_DIR}/include/harfbuzz" PARENT_SCOPE)

if(HARFBUZZ_LIBRARY_RELEASE)
  set(HARFBUZZ_FOUND TRUE PARENT_SCOPE)
else()
  set(HARFBUZZ_FOUND FALSE PARENT_SCOPE)
endif()
```

This seems to be a regression: these lines were introduced 26 days ago by https://github.com/microsoft/vcpkg/commit/401578455faece3db28615775420fb0f22f9bd07, I didn't have the warning before. The relevant sub-commits in the pull request are:
- https://github.com/microsoft/vcpkg/pull/22801/commits/0325a00cc84cc6a253f8583f4054e00d853dc4a0
- https://github.com/microsoft/vcpkg/pull/22801/commits/2fac3f8bb03c324616dc2cf246d080be28b53a04
- https://github.com/microsoft/vcpkg/pull/22801/commits/c8b1002f67113223b54253bb500aca8dc6a58ae8

If I remove all the `PARENT_SCOPE` in `C:/Users/dalbo/vcpkg/installed/x64-windows/share/harfbuzz/harfbuzzConfig.cmake`, then everything works as expected (no warnings + the variables are set):

```
PS C:\Users\dalbo\test\harfbuzz-warning\build4> cmake .. -DCMAKE_TOOLCHAIN_FILE="$HOME\vcpkg\scripts\buildsystems\vcpkg.cmake"
-- Building for: Visual Studio 16 2019
-- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.22000.
-- The C compiler identification is MSVC 19.29.30140.0
-- The CXX compiler identification is MSVC 19.29.30140.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found ZLIB: optimized;C:/Users/dalbo/vcpkg/installed/x64-windows/lib/zlib.lib;debug;C:/Users/dalbo/vcpkg/installed/x64-windows/debug/lib/zlibd.lib (found version "1.2.11")
-- HARFBUZZ_LIBRARY      = harfbuzz::harfbuzz
-- HARFBUZZ_LIBRARIES    = harfbuzz::harfbuzz
-- HARFBUZZ_INCLUDE_DIR  = C:/Users/dalbo/vcpkg/installed/x64-windows/include/harfbuzz
-- HARFBUZZ_INCLUDE_DIRS = C:/Users/dalbo/vcpkg/installed/x64-windows/include/harfbuzz
-- HARFBUZZ_FOUND        = TRUE
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/dalbo/test/harfbuzz-warning/build4
```

Hence, my intuition would be to fix this by either removing all the PARENT_SCOPE keywords, or alternatively using cache variables instead.

But strangely, these options seemed to have been discarded during a [discussion in the code review](https://github.com/microsoft/vcpkg/pull/22801/commits/0325a00cc84cc6a253f8583f4054e00d853dc4a0#r794506748). See for example this message by [@dg0yt](https://github.com/dg0yt):

```
Unlike Find modules, config files operate in separate scope. So they can't pollute the calling scope by accident. And must use PARENT_SCOPE for intentional changes.

Cache variables are normally used to save results from expensive system introspection or directly from the user. They are in global scope.
<Pkg>_LIBRARIES normally isn't a cache variable but regular (output) variable set by a Find module.
<Pkg>_LIBRARY usually is a cache variable (via find_library and as an input variable) for config-unaware Find modules. Config-aware Find modules derive it as a regular variable from <Pkg>_LIBRARY_RELEASE and <Pkg>_LIBRARY_DEBUG cache variables.
For completeness, <Pkg>_FOUND is a regular (output) variable, too.
```

But evidently, the initial assumption (`Unlike Find modules, config files operate in separate scope`) is not true, as per the warning message and behavior I get. It's not just on my machine: both me and my co-worker have the same warning on a clean install of VS 2019, cmake, and vcpkg.

So who's right? Should we use PARENT_SCOPE or not? Or use cache variables? Did anything else change more recently in vcpkg creating the warning? (I don't think they would have missed the warning, so I assume there was no warning when they tested the change?)

### PR DESCRIPTION
**Pull request by LilyWangLL**
Fixes #23694
Remove `PARENT_SCOPE` in harfbuzzConfig.cmake.in, because it will make the variables of `HARFBUZZ_LIBRARY` and so on can not set in CMake project. Like the following error:
```
Cannot set "HARFBUZZ_LIBRARY": current scope has no parent.
Cannot set "HARFBUZZ_LIBRARIES": current scope has no parent.
Cannot set "HARFBUZZ_INCLUDE_DIR": current scope has no parent.
Cannot set "HARFBUZZ_INCLUDE_DIRS": current scope has no parent.
Cannot set "HARFBUZZ_FOUND": current scope has no parent.
```

### CODE REVIEW

**Comment by LilyWangLL** [comment]
cc @cenit @dg0yt

**Review by JackBoosY — APPROVED** [review]

**Comment by dalboris** [comment]
Thanks a lot @LilyWangLL / @JackBoosY , these changes look good to me (bug reporter). I just hope removing PARENT_SCOPE isn't going to break someone else's workflow, I'm most likely not familiar with all possible use cases.

**Comment by dalboris** [comment]
@Neumann-A Thanks for the clarification.

**Comment by Neumann-A** [comment]
@LilyWangLL merge with master again?

File: ports/harfbuzz/harfbuzzConfig.cmake.in [hunk 1]
```diff
@@ -59,13 +59,19 @@ endif()
 get_filename_component(_INSTALL_DIR "${CMAKE_CURRENT_LIST_DIR}/../../" ABSOLUTE)
 target_include_directories(harfbuzz INTERFACE "${_INSTALL_DIR}/include/harfbuzz")
 
+set(HARFBUZZ_LIBRARY harfbuzz::harfbuzz)
```
Comment (LilyWangLL) [inline ports/harfbuzz/harfbuzzConfig.cmake.in:62]: Using `PARENT_SCOPE` will pass the variable to the upstream, and this port cannot use this variable, so add a variable to this port for itself.

**Comment by Neumann-A** [comment]
You should revert [cf01dcf](https://github.com/microsoft/vcpkg/pull/23709/commits/cf01dcf18abcafd1e960874f14f6b77a0c40a197)

`PARENT_SCOPE` in the wrapper is wrong!

**Comment by LilyWangLL** [comment]
> You should revert [cf01dcf](https://github.com/microsoft/vcpkg/pull/23709/commits/cf01dcf18abcafd1e960874f14f6b77a0c40a197)
> 
> `PARENT_SCOPE` in the wrapper is wrong!

But if doesn't use `PARENT_SCOPE`, other ports will be affected. Like `openmvs` `opencv4`.
cc @cenit @dg0yt 
`PARENT_SCOPE` and these codes added by commit https://github.com/microsoft/vcpkg/commit/401578455faece3db28615775420fb0f22f9bd07.
```
set(HARFBUZZ_LIBRARY harfbuzz::harfbuzz PARENT_SCOPE)
set(HARFBUZZ_LIBRARIES harfbuzz::harfbuzz PARENT_SCOPE)
set(HARFBUZZ_INCLUDE_DIR "${_INSTALL_DIR}/include/harfbuzz" PARENT_SCOPE)
set(HARFBUZZ_INCLUDE_DIRS "${_INSTALL_DIR}/include/harfbuzz" PARENT_SCOPE)

if(HARFBUZZ_LIBRARY_RELEASE)
  set(HARFBUZZ_FOUND TRUE PARENT_SCOPE)
else()
  set(HARFBUZZ_FOUND FALSE PARENT_SCOPE)
endif()
```

**Comment by Neumann-A** [comment]
> But if doesn't use PARENT_SCOPE, other ports will be affected. Like openmvs opencv4.

Then those ports are doing it wrong and need fixing

**Comment by JackBoosY** [comment]
> > But if doesn't use PARENT_SCOPE, other ports will be affected. Like openmvs opencv4.
> 
> Then those ports are doing it wrong and need fixing

If the config file only provides targets, we should probably add these macros to the wrapper in order to adapt to the MODULE mode.
Clearly, however, the hierarchy creates a problem here.

**Comment by Neumann-A** [comment]
> If the config file only provides targets, we should probably add these macros to the wrapper in order to adapt to the MODULE mode.

Wait are some of those variables `CACHE` variables in `MODULE` mode ? If yes set them as `CACHE` variables !

**Comment by dalboris** [comment]
Thanks everyone for working on this!

I'd just like to remind that this PR is supposed to fix the following bug: https://github.com/microsoft/vcpkg/issues/23694

The bug is that calling `find_package(harfbuzz)` in any CMakeLists produces the following warnings:

```
Cannot set "HARFBUZZ_LIBRARY": current scope has no parent.
Cannot set "HARFBUZZ_LIBRARIES": current scope has no parent.
Cannot set "HARFBUZZ_INCLUDE_DIR": current scope has no parent.
Cannot set "HARFBUZZ_INCLUDE_DIRS": current scope has no parent.
Cannot set "HARFBUZZ_FOUND": current scope has no parent.
```

because of all the lines in harfbuzzConfig.cmake.in which set a variable using PARENT_SCOPE, where in fact there is no parent scope. As of now, the PR isn't modifying these lines. It is just *adding* other lines, so the warning will remain and the bug will not be fixed.

I see a couple of options:
1. Remove all `PARENT_SCOPE` in harfbuzzConfig.cmake.in (original commit of this PR), and fix opencv and other clients so that they do not need the variables to be set at parent scope. This seems like the ideal fix but maybe not easy, I don't know.
2. As a workaround, in harfbuzzConfig.cmake.in, test whether there is a parent scope, and use PARENT_SCOPE or not accordingly. This seems fragile... what of use cases where there do exist a parent scope, but the client still expects the variables to be set at current scope? A possibility might be that when there is a parent scope, then the variables are set both as parent scope and current scope. And when there is no parent scope, there are set at current scope only.
3. As a workaround, in harfbuzzConfig.cmake.in, set the variables as cache variables instead of normal variables. This seems like the easiest solution to implement, but perhaps a bit too invasive and not ideal.

Thanks!

**Comment by dg0yt** [comment]
> If the config file only provides targets, we should probably add these macros to the wrapper in order to adapt to the MODULE mode. Clearly, however, the hierarchy creates a problem here.

There is no official Find module for Harfbuzz, so there cannot be a wrapper.

And going one step further, if there is no official Find module and upstream doesn't provide these variables in exported config, then the variables shouldn't be patched in by vcpkg IMHO. This is the wrong direction of dependencies, and it is inviting user dependencies on non-standard variables.
Instead, set variables as needed in consuming ports. 
And if you need an `ABC_FOUND` for `abc` config, there `find_package(ABC NAMES abc)`.

**Comment by dalboris** [comment]
Since implementing the "most correct" solution (fixing ports depending on Harfbuzz) seems to not get any traction, I would suggest in the meantime to implement the following workaround: testing if the current scope has a parent and applying PARENT_SCOPE or not accordingly. It would look like this:

```
# Set Harbuzz variables, either at the current scope or parent scope depending on the existence of
# a parent scope. Ideally, we should never set the variable at parent scope, but it is a workaround
# for some ports that expect this behavior. See https://github.com/microsoft/vcpkg/issues/23694
get_directory_property(hasParent PARENT_DIRECTORY)
if(hasParent)
    set(HARFBUZZ_LIBRARIES harfbuzz::harfbuzz PARENT_SCOPE)
    set(HARFBUZZ_INCLUDE_DIR "${_INSTALL_DIR}/include/harfbuzz" PARENT_SCOPE)
    set(HARFBUZZ_INCLUDE_DIRS "${_INSTALL_DIR}/include/harfbuzz" PARENT_SCOPE)
    if(HARFBUZZ_LIBRARY_RELEASE)
        set(HARFBUZZ_FOUND TRUE PARENT_SCOPE)
    else()
        set(HARFBUZZ_FOUND FALSE PARENT_SCOPE)
    endif()
else()
    set(HARFBUZZ_LIBRARIES harfbuzz::harfbuzz)
    set(HARFBUZZ_INCLUDE_DIR "${_INSTALL_DIR}/include/harfbuzz")
    set(HARFBUZZ_INCLUDE_DIRS "${_INSTALL_DIR}/include/harfbuzz")
    if(HARFBUZZ_LIBRARY_RELEASE)
        set(HARFBUZZ_FOUND TRUE)
    else()
        set(HARFBUZZ_FOUND FALSE)
    endif()
endif()
```

Should I make a PR with such change? Or maybe @LilyWangLL you could integrate this change to this PR so we can see if the checks pass with this approach?

Thanks!

**Comment by dalboris** [comment]
Any updates on this?

**Comment by LilyWangLL** [comment]
> Any updates on this?

Sorry I missed the previous comment, I will test this soon. Thanks very much for your suggestion~

**Comment by dalboris** [comment]
@LilyWangLL Perfect, thanks a lot!

**Comment by LilyWangLL** [comment]
Closed as duplicate to #25092

**Pull request closed without merging**